presentation/src/main/java/org/sagebionetworks/research/presentation/speech/TextToSpeechService.kt

```diff
@@ -0,0 +1,437 @@
+        }
+
+        return super.onUnbind(intent)
+    }
+
+    override fun onDestroy() {
+        textToSpeech?.stop()
+        textToSpeech?.shutdown()
+        textToSpeech = null
+        cleanupMediator.removeObserver(cleanupObserver)
+        clear()
+        super.onDestroy()
+    }
+
+    override fun onInit(status: Int) {
+        if (status == TextToSpeech.SUCCESS) {
+            val languageAvailable = textToSpeech!!.isLanguageAvailable(Locale.getDefault())
+            // >= 0 means LANG_AVAILABLE, LANG_COUNTRY_AVAILABLE, or LANG_COUNTRY_VAR_AVAILABLE
+            if (languageAvailable >= 0) {
+                textToSpeech!!.language = Locale.getDefault()
+                textToSpeech!!.setOnUtteranceProgressListener(object : UtteranceProgressListener() {
+                    override fun onDone(text: String?) {
+                        val tts = textToSpeech
+                        if (tts != null && !tts.isSpeaking) {
+                            val data = futureSpeechData
+                            if (data == null || !data.hasQueuedSpeech()) {
+                                // The TTS has just finished so we can shut it down and set the state to idle
+                                _state.postValue(TextToSpeechState(IDLE, null))
+                                textToSpeech!!.shutdown()
+                            } else {
+                                // More speech is queued
+                                _state.postValue(TextToSpeechState(QUEUED, null))
+                            }
+                        }
+                    }
+
+                    override fun onError(text: String?) {
+                        _state.postValue(TextToSpeechState(ERROR, null))
+                    }
+
+                    override fun onStart(text: String?) {
+                        _state.postValue(TextToSpeechState(SPEAKING, text))
+                    }
+                })
+                if (textToSpeakOnInit != null) {
+                    speakText(textToSpeakOnInit!!)
+                }
+            } else {
+                LOGGER.warn("Language not available for TTS")
+                textToSpeech?.shutdown()
+                textToSpeech = null
+            }
+        } else {
+            LOGGER.warn("Failed to initialize TTS with error code $status")
+            textToSpeech = null
+        }
+    }
+
+    /**
+     * Register the given speech map to be spoken at the times determined by the given countdown and duration.
+     * @param duration the duration of the current step
+     * @param countdown the countdown LiveData for the current step
+     * @param speechMap the map where keys are offsets into the countdown and values are instructions that should be
+     * spoken to the user.
+     */
+    fun registerSpeechesOnCountdown(duration: Duration, countdown: LiveData<Long>, speechMap: Map<String, String>) {
+        if (LOGGER.isDebugEnabled) {
```

Comment:
Should I be checking if isDebugEnabled before every log statement? 

If so, it seems odd to have to do that extra line, and we could add an extension function to use instead of .debug() something like this...

```
fun Logger.debugChecked(log: String?) {
   if (this.isDebugEnabled) {
    this.debug(log)
  }
}
```
Then you can just call LOGGER.debugChecked() everywhere that extension is imported

Reply:
Oops it looks as though this check isn't needed. I can't remember where I first saw it but I'll remove it from this pr.

Reply:
This is mostly for when the operations to generate the String is expensive, e.g. complex String building. If you want an extension function I think it would need to take a Function<String> rather than String.